src/NuGet.Core/NuGet.Commands/RestoreCommand/Utility/BuildAssetsUtils.cs

```diff
@@ -71,50 +74,96 @@ public static class BuildAssetsUtils
+                // Add additional conditionals for cross targeting
+                var isCrossTargeting = request.Project.RestoreMetadata.CrossTargeting
+                    || request.Project.TargetFrameworks.Count > 1;
 
-                // Add targets/props under the condition they need
-                props.Add(CrossTargetingCondition, crossTargetingAssets.Props);
-                targets.Add(CrossTargetingCondition, crossTargetingAssets.Targets);
+                Debug.Assert((!request.Project.RestoreMetadata.CrossTargeting && (request.Project.TargetFrameworks.Count < 2)
+                    || (request.Project.RestoreMetadata.CrossTargeting)),
+                    "Invalid crosstargeting and framework count combination");
+
+                if (isCrossTargeting)
+                {
+                    // Find all global targets from buildCrossTargeting
+                    var crossTargetingAssets = GetTargetsAndPropsForCrossTargeting(
+                            targetGraphs,
+                            repositories,
+                            context,
+                            request,
+                            includeFlagGraphs);
+
+                    var crossProps = new MSBuildRestoreImportGroup();
+                    crossProps.Position = 0;
+                    crossProps.Conditions.Add(CrossTargetingCondition);
+                    crossProps.Imports.AddRange(crossTargetingAssets.Props);
+                    props.Add(crossProps);
+
+                    var crossTargets = new MSBuildRestoreImportGroup();
+                    crossTargets.Position = 0;
+                    crossTargets.Conditions.Add(CrossTargetingCondition);
+                    crossTargets.Imports.AddRange(crossTargetingAssets.Targets);
+                    targets.Add(crossTargets);
+                }
 
                 // Find TFM specific assets from the build folder
                 foreach (var pair in buildAssetsByFramework)
                 {
                     // There could be multiple string matches
                     foreach (var match in GetMatchingFrameworkStrings(project, pair.Key))
                     {
+                        var frameworkCondition = string.Format(CultureInfo.InvariantCulture, TargetFrameworkCondition, match);
+
                         // Add entries regardless of if imports exist,
                         // this is needed to trigger conditionals
-                        if (!props.ContainsKey(match))
+                        var propsGroup = new MSBuildRestoreImportGroup();
+
+                        if (isCrossTargeting)
                         {
-                            props.Add(match, pair.Value.Props);
+                            propsGroup.Conditions.Add(frameworkCondition);
                         }
 
-                        if (!targets.ContainsKey(match))
+                        propsGroup.Imports.AddRange(pair.Value.Props);
+                        propsGroup.Position = 2;
```

Comment:
I wondered why 2.  I think it just needs to be after position 0 (added above), but why not 1?  Consider using a local (e.g.:  `position + 1`).